api/v1alpha1/nicclusterpolicy_webhook.go

```diff
@@ -0,0 +1,305 @@
+	nicClusterPolicyLog.Info("validate update", "name", w.Name)
+	return w.validateNicClusterPolicy()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (w *NicClusterPolicy) ValidateDelete() error {
+	nicClusterPolicyLog.Info("validate delete", "name", w.Name)
+
+	// Validation for delete call is not required
+	return nil
+}
+
+/*
+We are validating here NicClusterPolicy:
+ 1. IBKubernetes.pKeyGUIDPoolRangeStart and IBKubernetes.pKeyGUIDPoolRangeEnd must be valid GUID and valid range.
+ 2. OFEDDriver.version must be a valid ofed version.
+ 3. RdmaSharedDevicePlugin.Config.
+    3.1. Configuration is a valid JSON and check its schema.
+    3.2. resourceName is valid for k8s.
+    3.3. At least one of the supported selectors exists.
+    3.4. All selectors are strings.
+ 4. SriovNetworkDevicePlugin.Config.
+    4.1. Configuration is a valid JSON and check its schema.
+    4.2. resourceName is valid for k8s.
+    4.3. At least one of the supported selectors exists.
+    4.4. All selectors are strings.
+*/
+func (w *NicClusterPolicy) validateNicClusterPolicy() error {
+	var allErrs field.ErrorList
+	// Validate IBKubernetes
+	ibKubernetes := w.Spec.IBKubernetes
+	if ibKubernetes != nil {
+		allErrs = append(allErrs, ibKubernetes.validate(field.NewPath("spec").Child("ibKubernetes"))...)
+	}
+
+	// Validate OFEDDriverSpec
+	ofedDriver := w.Spec.OFEDDriver
+	if ofedDriver != nil {
+		allErrs = append(allErrs, ofedDriver.validateVersion(field.NewPath("spec").Child("ofedDriver"))...)
+	}
+	// Validate RdmaSharedDevicePlugin
+	rdmaSharedDevicePlugin := w.Spec.RdmaSharedDevicePlugin
+	if rdmaSharedDevicePlugin != nil {
+		allErrs = append(allErrs, w.Spec.RdmaSharedDevicePlugin.validateRdmaSharedDevicePlugin(
+			field.NewPath("spec").Child("rdmaSharedDevicePlugin"))...)
+	}
+	// Validate SriovDevicePlugin
+	sriovNetworkDevicePlugin := w.Spec.SriovDevicePlugin
+	if sriovNetworkDevicePlugin != nil {
+		allErrs = append(allErrs, w.Spec.SriovDevicePlugin.validateSriovNetworkDevicePlugin(
+			field.NewPath("spec").Child("sriovNetworkDevicePlugin"))...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "mellanox.com", Kind: "NicClusterPolicy"},
+		w.Name, allErrs)
+}
+func (dp *DevicePluginSpec) validateSriovNetworkDevicePlugin(fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	var sriovNetworkDevicePluginConfigJSON map[string]interface{}
+	sriovNetworkDevicePluginConfig := *dp.Config
+
+	// Validate if the SRIOV Network Device Plugin Config is a valid json
+	if err := json.Unmarshal([]byte(sriovNetworkDevicePluginConfig), &sriovNetworkDevicePluginConfigJSON); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("Config"), dp.Config,
+			"Invalid json of SriovNetworkDevicePluginConfig"))
+		return allErrs
+	}
+
+	// Load the JSON Schema
+	sriovNetworkDevicePluginSchemaLoader := gojsonschema.NewReferenceLoader(
```

Comment:
can you load all schema once, then use them for validation ?

this can be done using the Schema obj

https://github.com/xeipuuv/gojsonschema/blob/b076d39a02e5015af0a2a96636e4cc479ecd9f45/schema.go#L49

currently on each request you are reading the same schema information from the file system.

Reply:
done